// backend/dist/services/dynamodb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.connectionsService = exports.cardsService = void 0;
const client_dynamodb_1 = require("@aws-sdk/client-dynamodb");
const lib_dynamodb_1 = require("@aws-sdk/lib-dynamodb");
const client = new client_dynamodb_1.DynamoDBClient({});
const docClient = lib_dynamodb_1.DynamoDBDocumentClient.from(client);
const CARDS_TABLE = process.env.CARDS_TABLE;
const CONNECTIONS_TABLE = process.env.CONNECTIONS_TABLE;
exports.cardsService = {
    async create(card) {
        await docClient.send(new lib_dynamodb_1.PutCommand({
            TableName: CARDS_TABLE,
            Item: card,
        }));
        return card;
    },
    async get(id) {
        const result = await docClient.send(new lib_dynamodb_1.GetCommand({
            TableName: CARDS_TABLE,
            Key: { id },
        }));
        return result.Item || null;
    },
    async update(id, updates) {
        const updateExpressions = [];
        const expressionAttributeNames = {};
        const expressionAttributeValues = {};
        Object.entries(updates).forEach(([key, value]) => {
            if (value !== undefined) {
                updateExpressions.push(`#${key} = :${key}`);
                expressionAttributeNames[`#${key}`] = key;
                expressionAttributeValues[`:${key}`] = value;
            }
        });
        updateExpressions.push('#updatedAt = :updatedAt');
        expressionAttributeNames['#updatedAt'] = 'updatedAt';
        expressionAttributeValues[':updatedAt'] = new Date().toISOString();
        const result = await docClient.send(new lib_dynamodb_1.UpdateCommand({
            TableName: CARDS_TABLE,
            Key: { id },
            UpdateExpression: `SET ${updateExpressions.join(', ')}`,
            ExpressionAttributeNames: expressionAttributeNames,
            ExpressionAttributeValues: expressionAttributeValues,
            ReturnValues: 'ALL_NEW',
        }));
        return result.Attributes;
    },
    async delete(id) {
        await docClient.send(new lib_dynamodb_1.DeleteCommand({
            TableName: CARDS_TABLE,
            Key: { id },
        }));
    },
    async list() {
        const result = await docClient.send(new lib_dynamodb_1.ScanCommand({
            TableName: CARDS_TABLE,
        }));
        return result.Items || [];
    },
    async listByColumn(column) {
        const result = await docClient.send(new lib_dynamodb_1.QueryCommand({
            TableName: CARDS_TABLE,
            IndexName: 'ColumnIndex',
            KeyConditionExpression: '#column = :column',
            ExpressionAttributeNames: { '#column': 'column' },
            ExpressionAttributeValues: { ':column': column },
        }));
        return result.Items || [];
    },
};
exports.connectionsService = {
    async add(connectionId) {
        await docClient.send(new lib_dynamodb_1.PutCommand({
            TableName: CONNECTIONS_TABLE,
            Item: {
                connectionId,
                connectedAt: new Date().toISOString(),
                ttl: Math.floor(Date.now() / 1000) + 86400, // 24 hours
            },
        }));
    },
    async remove(connectionId) {
        await docClient.send(new lib_dynamodb_1.DeleteCommand({
            TableName: CONNECTIONS_TABLE,
            Key: { connectionId },
        }));
    },
    async list() {
        const result = await docClient.send(new lib_dynamodb_1.ScanCommand({
            TableName: CONNECTIONS_TABLE,
        }));
        return (result.Items || []).map(item => item.connectionId);
    },
};
