// backend/dist/handlers/ai-task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = void 0;
const uuid_1 = require("uuid");
const bedrock_1 = require("../services/bedrock");
const dynamodb_1 = require("../services/dynamodb");
const websocket_1 = require("../services/websocket");
const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET,POST,PUT,DELETE,OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
};
const handler = async (event) => {
    console.log('AI Task Creation:', event.body);
    // Handle OPTIONS for CORS
    if (event.httpMethod === 'OPTIONS') {
        return { statusCode: 200, headers, body: '' };
    }
    try {
        const { description } = JSON.parse(event.body || '{}');
        if (!description) {
            return {
                statusCode: 400,
                headers,
                body: JSON.stringify({ error: 'Description required' }),
            };
        }
        // Get current board context
        const cards = await dynamodb_1.cardsService.list();
        const boardContext = {
            totalCards: cards.length,
            columns: ['To Do', 'In Progress', 'Done'],
        };
        // Call Bedrock
        const suggestion = await (0, bedrock_1.invokeClaudeForTask)(description, boardContext);
        // Create card from AI suggestion
        const card = {
            id: (0, uuid_1.v4)(),
            title: suggestion.title,
            description: suggestion.description,
            column: 'To Do',
            position: 0,
            storyPoints: suggestion.storyPoints,
            priority: suggestion.priority,
            acceptanceCriteria: suggestion.acceptanceCriteria,
            aiGenerated: true,
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString(),
        };
        await dynamodb_1.cardsService.create(card);
        // Broadcast to all clients
        await (0, websocket_1.broadcastToAll)({
            type: 'card_created',
            data: card,
        });
        return {
            statusCode: 201,
            headers,
            body: JSON.stringify(card),
        };
    }
    catch (error) {
        console.error('AI Task Error:', error);
        return {
            statusCode: 500,
            headers,
            body: JSON.stringify({
                error: 'Failed to create AI task',
                message: error.message
            }),
        };
    }
};
exports.handler = handler;
