// backend/dist/handlers/cards.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = void 0;
const uuid_1 = require("uuid");
const dynamodb_1 = require("../services/dynamodb");
const websocket_1 = require("../services/websocket");
const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET,POST,PUT,DELETE,OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
};
const handler = async (event) => {
    console.log('Event:', JSON.stringify(event, null, 2));
    try {
        const method = event.httpMethod;
        const path = event.path;
        // OPTIONS for CORS
        if (method === 'OPTIONS') {
            return { statusCode: 200, headers, body: '' };
        }
        // GET /cards - List all cards
        if (method === 'GET' && path === '/cards') {
            const cards = await dynamodb_1.cardsService.list();
            return {
                statusCode: 200,
                headers,
                body: JSON.stringify(cards),
            };
        }
        // GET /cards/{id} - Get single card
        if (method === 'GET' && path.startsWith('/cards/')) {
            const id = path.split('/')[2];
            const card = await dynamodb_1.cardsService.get(id);
            if (!card) {
                return {
                    statusCode: 404,
                    headers,
                    body: JSON.stringify({ error: 'Card not found' }),
                };
            }
            return {
                statusCode: 200,
                headers,
                body: JSON.stringify(card),
            };
        }
        // POST /cards - Create card
        if (method === 'POST' && path === '/cards') {
            const request = JSON.parse(event.body || '{}');
            const card = {
                id: (0, uuid_1.v4)(),
                title: request.title,
                description: request.description,
                column: request.column,
                position: request.position || 0,
                storyPoints: request.storyPoints,
                priority: request.priority,
                acceptanceCriteria: request.acceptanceCriteria,
                aiGenerated: false,
                createdAt: new Date().toISOString(),
                updatedAt: new Date().toISOString(),
            };
            await dynamodb_1.cardsService.create(card);
            // Broadcast to all connected clients
            await (0, websocket_1.broadcastToAll)({
                type: 'card_created',
                data: card,
            });
            return {
                statusCode: 201,
                headers,
                body: JSON.stringify(card),
            };
        }
        // PUT /cards/{id} - Update card
        if (method === 'PUT' && path.startsWith('/cards/')) {
            const id = path.split('/')[2];
            const updates = JSON.parse(event.body || '{}');
            const card = await dynamodb_1.cardsService.update(id, updates);
            // Broadcast to all connected clients
            await (0, websocket_1.broadcastToAll)({
                type: 'card_updated',
                data: card,
            });
            return {
                statusCode: 200,
                headers,
                body: JSON.stringify(card),
            };
        }
        // DELETE /cards/{id} - Delete card
        if (method === 'DELETE' && path.startsWith('/cards/')) {
            const id = path.split('/')[2];
            await dynamodb_1.cardsService.delete(id);
            // Broadcast to all connected clients
            await (0, websocket_1.broadcastToAll)({
                type: 'card_deleted',
                data: { id },
            });
            return {
                statusCode: 204,
                headers,
                body: '',
            };
        }
        return {
            statusCode: 404,
            headers,
            body: JSON.stringify({ error: 'Not found' }),
        };
    }
    catch (error) {
        console.error('Error:', error);
        return {
            statusCode: 500,
            headers,
            body: JSON.stringify({ error: 'Internal server error' }),
        };
    }
};
exports.handler = handler;
